edit Overview text (#665)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

# Overview {#overview}

The protocol is executed by a large set of Clients and a pair of Aggregators.
Each Client's input to the protocol is its measurement (or set of measurements,
e.g., counts of some user behavior). Given a set of measurements
`meas_1, ..., meas_N` held by the Clients, and an "aggregation parameter"
`agg_param` shared by the Aggregators, the goal of DAP is to compute
`agg_result = F(agg_param, meas_1, ..., meas_N)` for some function `F` while
revealing nothing else about the measurements. We call `F` the "aggregation
function" and `agg_result` the "aggregate result".

DAP is extensible in that it allows for the addition of new cryptographic
schemes that compute different aggregation functions, determined by the
Verifiable Distributed Aggregation Function, or

> {{!VDAF=I-D.draft-irtf-cfrg-vdaf-14}}, used to compute it.

Reviewer: ```suggestion
{{!VDAF=I-D.draft-irtf-cfrg-vdaf-14}}, used in the computation.
```
?

Author: I don't disagree with your phrasing but am not sure why it's better. Is the idea to allow for things besides the VDAF to be involved in the computation?

Reviewer 2: I prefer the current verbiage: a VDAF computes an aggregation function, it's not just one part of the computation.

  measurement from which it was generated.

* Aggregators can compute secret shares of the aggregate result by aggregating
  their shares locally into "aggregate shares", which may later be merged into
  the aggregate result.

DAP is not compatible with all VDAFs. Some VDAFs allow measurements to be
aggregated multiple times with a different aggregation parameter. DAP only

> allows each measurement to be aggregated once. Some VDAFs produce aggregate

Reviewer 2: At first I read this as "DAP doesn't support Mastic", which I don't think is the intent. It's true that there are certain things we can do with Poplar1/Mastic that we can't do in DAP, but there are still useful applications of Mastic (and Poplar1, kind of) that DAP supports.

Basically I would just avoid conflating multi-collection and order-sensitive VDAFs here.

  measurement from which it was generated.

* Aggregators can compute secret shares of the aggregate result by aggregating
  their shares locally into "aggregate shares", which may later be merged into
  the aggregate result.

> DAP is not compatible with all VDAFs. Some VDAFs produce aggregate
> values which depend on the order in which the measurements are aggregated. DAP

Reviewer 2: might: This is situation is very hypothetical
values -> results: consistency with terminology
```suggestion
DAP is not compatible with all VDAFs. Some VDAFs might produce aggregate
results which depend on the order in which the measurements are aggregated. DAP
```

{{dap-topology}} illustrates which participants exchange HTTP messages. Arrows
go from HTTP clients to HTTP servers. Note that some DAP participants may be
HTTP clients sometimes but HTTP servers at other times. It is even possible for
a single entity to perform multiple DAP roles. For example, the Collector could
also be one of the Aggregators.

In the course of a measurement task ({{task-configuration}}), each Client
records its own measurement, packages it up into a report, and sends it to the
Leader. Each share is encrypted to only one of the two Aggregators so that even

> though they pass through the Leader, the Leader is unable to see or modify the

Reviewer 2: ```suggestion
though both pass through the Leader, the Leader is unable to see or modify the
```

Helper's share. Depending on the task, the Client may only send one report or
may send many reports over time.

The Leader distributes the shares to the Helper and orchestrates the process of
verifying them and assembling them into a aggregate shares for the Collector.
Depending on the VDAF, it may be possible to process each report as it is

> uploaded, or it may be necessary to wait until the Collector transmits a
> collection request before processing can begin.

Reviewer 2: WDYT?
```suggestion
uploaded, or it may be necessary to wait until the Collector initializes a collection job before processing can begin.
```

Author: Sure, that's more consistent with how we discuss collection jobs elsewhere.

## Validating Measurements {#validating-inputs}

An essential goal of any data collection pipeline is ensuring that the data
being aggregated is "valid". For example, each measurement might be expected to
be a number between 0 and 10. In DAP, input validation is complicated by the
fact that none of the entities other than the Client ever sees that Client's
plaintext measurement. To an Aggregator, a secret share of a valid measurement
is indistinguishable from a secret share of an invalid measurement.

DAP validates inputs using an interactive computation between the Leader and
Helper. At the beginning of this computation, each Aggregator holds an input
share uploaded by the Client. At the end of the computation, each Aggregator

> either holds an output share that is ready to be aggregated or rejects the

Reviewer 3: Having an active and passive verb on either side of this "either ... or" feels unusual to me
```suggestion
either obtains an output share that is ready to be aggregated or rejects the
```

taken. For instance, if the task is measuring the latency of some operation, the
proof might demonstrate that the value reported was between 0 and 60 seconds.
But to report which of `N` options a user selected, the report might contain `N`
integers and the proof would demonstrate that `N-1` of them were `0` and the
other was `1`.

"Validity" is distinct from "correctness". For instance, the user might have
spent `30` seconds on a task but might report `60` seconds. This is a problem
with any measurement system and DAP does not attempt to address it. DAP merely
ensures that the data is within the chosen limits, so the Client could not
report `10^6`  or `-20` seconds.

## Replay Protection {#replay-protection}

Another goal of DAP is to mitigate replay attacks in which a report is
aggregated in multiple batches or multiple times in a single batch. This would
allow the attacker to learn more information about the underlying measurement
than it would otherwise.

When a Client generates a report, it also generates a random nonce, called the
"report ID". Each Aggregator is responsible for storing the IDs of reports it

> has aggregated and rejecting replayed reports.

Reviewer: Should we say `repeated` instead of `replayed`, to be more encompassing of errors rather than malicious action?

Author: I don't think "repeated" would be wrong, but we frequently discuss replay attacks and replayed reports elsewhere in the draft, so I think consistency is important.
